Structured Health Reporter + EndpointManager Modular Health Checks

tommyp1ckles opened the pull request:
Adds "structured health reporter" which builds on top of existing cell.HealthReporter type. 

The intention is to provide a mechanism to do health status reporting on more complex modules, namely ones with many sub-components that all need to be health checked as well. 

Subsequent commits in this pull request use the structured reporter to create health status reporting for the endpointmanager module to provide better visibility into the runtime status of endpointmanager/endpoints in a Cilium Agent.

The structured health reporter uses a tree structured to aggregate sub-reporters up to a top level report, which is ultimately passed to the "module" scoped health reporter to declare the status of the module component.

This PR adds the structured health reporter, and in subsequent commits implements structured health reporting on the endpointmanager module to provide better insight in the system health of Endpoints/EndpointManager.

```
> cilium status --verbose

  ...
Modules Health:
  Module             Status    Message                  Last Updated
  daemon             Unknown   No status reported yet            n/a
  endpoint-manager   OK                  9s
  
   cilium-endpoint-1058
    cep-k8s-sync        OK sync-to-k8s-ciliumendpoint (1058) 3m10.022180274s ago (x20)
    datapath-regenerate OK Endpoint regeneration successful  3m10.016021308s ago (x2)
    policy map sync     OK sync-policymap-1058               3m6.639716331s ago  (x5)
   cilium-endpoint-1488
    cep-k8s-sync        OK sync-to-k8s-ciliumendpoint (1488) 3m10.022419627s ago (x20)
    datapath-regenerate OK Endpoint regeneration successful  3m10.014718641s ago (x2)
    policy map sync     OK sync-policymap-1488               3m6.633931551s ago  (x5)
   cilium-endpoint-1704
    cep-k8s-sync        OK sync-to-k8s-ciliumendpoint (1704) 3m10.022513188s ago (x20)
    datapath-regenerate OK Endpoint regeneration successful  3m10.014787702s ago (x2)
    policy map sync     OK sync-policymap-1704               3m6.632632525s ago  (x5)
   cilium-endpoint-2366
    cep-k8s-sync        OK sync-to-k8s-ciliumendpoint (2366) 3m10.023392989s ago (x20)
    datapath-regenerate OK Endpoint regeneration successful  3m10.021365103s ago (x1)
...
```

Follow up work:
* Add metrics

tommyp1ckles committed:
pkg/health: improve displaying of multiline health reports.

Health reports will often be multiple lines, this improves formatting in
such case.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
pkg/controller: Add parameter for cell.HealthReporter.

When passed, controller runs will automatically update the health
reporter.

Upon failure, a passed reporter will be degraded, otherwise it will be
marked as "ok" (i.e. healthy).

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
endpointmanager: add modular health checks to epm componenets.

Adds modular health checking to endpointmanager by using structured
reporter to health check the following reconciliation tasks:

* K8s endpoint sync
* Endpoint regeneration
* Endpoint policy map sync

This means that the endpointmanager will be reported as healthy if all
endpoints have their ep datapath regenerated, policy map in sync and cep
sync controller has reconciled successfully.

This will make it easier to understand the state of the
endpointmanager/endpoint components of Cilium - providing a way to
determine if there are any health problems with endpoints.

Example output:
```
  endpoint-manager   OK                  4s

  cilium-endpoint-1314
   datapath-regenerate OK Endpoint regeneration successful
   sync-policymap-1314 OK sync-policymap-1314
  cilium-endpoint-56
   sync-to-k8s-ciliumendpoint OK sync-to-k8s-ciliumendpoint (56)
   datapath-regenerate        OK Endpoint regeneration successful
   sync-policymap-56          OK sync-policymap-56
```

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

joamaki commented on pkg/hive/cell/structured.go:
```diff
@@ -0,0 +1,384 @@
+	"golang.org/x/exp/maps"
+)
+
+type children map[string]struct{}
```
btw, I've started using https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#Set instead of `map[Key]struct{}` recently as that provides a nice generic set API and we're anyway depending on the apimachinery package.

joamaki commented on pkg/hive/cell/structured.go:
```diff
@@ -0,0 +1,384 @@
+	return rootScope(hr)
+}
+
+// TODO: Add flush function.
```
Can we address this TODO in this PR?

joamaki commented on pkg/hive/cell/structured.go:
```diff
@@ -0,0 +1,384 @@
+	}
+}
+
+func getSubReporterV2(parent Scope, name string, isReporter bool) *subReporter {
```
Remove `V2`?

joamaki commented on pkg/hive/cell/structured.go:
```diff
@@ -0,0 +1,384 @@
+}
+
+func (s *subreporterBase) addChild(pid string, name string, isReporter bool) string {
+	id := name + "_" + uuid.New().String()
```
Wonder if a global counter would be nicer here. It would also inform in what order these were added which might be useful. And looks like we're sorting these later, so using a global counter would probably preserve the order across restarts. Do these ids need to be unique outside this process?

tommyp1ckles replied:
> Do these ids need to be unique outside this process?

For now, no.

Makes sense, I've changed this to just use a counter within the reporter. I've changed it to use a counter within the reporter base scope and prefix the ID with the uid, that should preserve order when realizing since we sort by IDs.

What would be the benefit of having this be global across all reporters?

joamaki commented on pkg/hive/cell/structured.go:
```diff
@@ -0,0 +1,384 @@
+
+func (s *subReporter) Degraded(message string, err error) {
+	var m string
+	// TODO: For now, we just append the error to the message, this won't be ideal
```
Can we address this TODO now? If not, perhaps doing a follow-up issue would make sense.

tommyp1ckles replied:
Yeah I'm kind of kicking this can down the road a bit 😅 

While working on this I went back and fourth on removing the error from the `Degraded` function.  I ended up keeping it to keep the PR smaller and since I'm still not entirely sure what the HealthReporter API should ultimately look like.

I think eventually we'll want to replace the passed status values with some kind of interface like:

```go
type Report interface {
    String() string
    JSON() []byte
    Errs() []error
}
```

(This will also be important since currently we send the status update as raw text so on the receiving end we have no good way to decode the structure).

In the interest of keeping this PR manageable, I'm going to defer this to a future PR. At the moment we don't even support JSON for the health status output so it doesn't make that much of a difference.

I'll address this in a follow up after these changes are merged.

joamaki commented on pkg/hive/cell/structured.go:
```diff
@@ -0,0 +1,384 @@
+// Example:
+//
+// 1. Init root scope (note: this is provided to modules automatically).
+// root := rootScope(ctx, hr)
```
Examples use `ctx` and the indentation for the examples seems to be a bit messed up.

joamaki commented on pkg/hive/cell/structured.go:
```diff
@@ -0,0 +1,384 @@
+
+// GetSubScope creates a new context with a sub reporter scope context.
+// This creates a new node in the structured health reporter tree, and any calls
+// to GetReporter from the returned context will return a reporter that is a child
```
Both this comment and the commit description talk about context, but looks like use of `Context` was dropped.

joamaki commented on pkg/hive/cell/structured.go:
```diff
@@ -0,0 +1,384 @@
+			r.base.hr.Degraded(statusTree.toString(2), nil)
+		}
+	}
+	t, err := trigger.NewTrigger(trigger.Parameters{
```
Do we shut down the trigger? e.g. will this leak a goroutine?

Might maybe be worth considering use of https://github.com/cilium/cilium/blob/main/pkg/hive/job/job.go#L339 along with `WithTrigger`? Would make it easier to manage the lifecycle of these.

OTOH, maybe we want to use health reporting with `pkg/hive/job`... :)

tommyp1ckles commented on pkg/hive/cell/lifecycle.go:
```diff
@@ -0,0 +1,24 @@
+// is cast into a regular context.Context.
+//
+// This should only be used by cell internals, other packages should use hive.Lifecycle instead.
+type Lifecycle interface {
```
@joamaki This doesn't feel great - but I couldn't find a better way to do this without refactoring a bunch of things. Not sure if you have any better ideas here 🤔

joamaki replied:
Let's try moving types from `pkg/hive/lifecycle.go` to `pkg/hive/cell/lifecycle.go` (without `DefaultLifecycle` though), and then adding type aliases to `pkg/hive/lifecycle.go` to keep everything working. This would also be nice in that we'll then only need to import `cell` package when creating new cells and not both `cell` and `hive`.

jrajahalme commented on pkg/controller/controller.go:
```diff
@@ -277,7 +280,10 @@ func (c *controller) runController(params ControllerParams) {
+				c.recordError(err, params.HealthReporter)
+				if params.HealthReporter != nil {
+					params.HealthReporter.Degraded(c.name, err)
+				}
```
This seems to be duplicated code, `recordError()` already calls `Degraded()`?

tommyp1ckles replied:
yup, fixed!

jrajahalme reviewed (approved):
LGTM apart from some apparently duplicated code.

tommyp1ckles committed:
hive: move lifecycle types to separate package.

Currently using lifecycle in cell is not possible as that would introduce a cyclic dependency.
This moves the type definition of HookContext to a separate package, as
well as adds separate Lifecycle/HookInterface definitions that can be
imported by pkg/hive/cell.

Finally, pkg/hive now injects a wrapped hive.Lifecycle that converts pkg/hive/cell/lifecycle hooks.
Thus allowing pkg/cell to depend on lifecycle hooks.

This is done to avoid having to refactor the lifecycle types in pkg/hive
which would result in having to fix imports across the codebase.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
hive: add structured health reporter.

Structured health reporter initializes a context.Context object with a
module scoped cell.HealthReporter.

This can be then used to propegate a structured health reporter tree,
where leaves represent active health reporters and intermediary nodes
provide a common scope for all sub reporters.

In subsequent commits, this will be used to provide modular health
status reporting for more complex modules such as endpoint manager where
many "sub-components" must also be healthy in order for the module to
report as healthy.

Currently, this is an opt-in layer built on top of the existing
cell.HealthReporter interface, by layering functionality on the existing
health reporter provided by the cell.Module.

Currently the context is initialized with the structured tree reporter at the module level.
To avoid the risk of deadlocks, the entire reporter tree uses one lock
which is initialized when calling WithStructuredReporter. This is shared
among all subreporters.

 Example:

 1. Init root scope (note: this is provided to modules automatically).
 root := rootScope(ctx, hr)

	root

 2. Create endpoint-manager subscope, and reporter under that scope (with ok!)
 endpointManagerScope := GetSubScope(ctx, root, "endpoint-manager")
 GetHealthReporter(endpointManagerScope, "endpoint-000").OK("it works!")

	 root(OK)
		└── scope(endpoint-manager, OK)
			└── reporter(endpoint-000, OK)

 3. Create another reporter under that scope with degraded
 GetHealthReporter(endpointManagerScope, "endpoint-000").Degraded("oh no!")

	 root(Degraded)
		└── scope(endpoint-manager, Degraded)
			└── reporter(endpoint-000, OK)
			└── reporter(endpoint-000, Degraded)

 4. Close the endpoint-manager scope
 endpointManagerScope.Stopped("no more status to report!")

	root(OK) // status has been reported, but we no longer have any degraded status
		 // default to ok status.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

joamaki commented on pkg/hive/cell/module.go:
```diff
@@ -56,16 +57,31 @@ func (m *module) logger(log logrus.FieldLogger) logrus.FieldLogger {
+}
+
+func (m *module) moduleReporterScope(p Health, lc lifecycle.Lifecycle) Scope {
+	rs := rootScope(p.forModule(m.id))
```
BTW, I was thinking we might want "fully-qualified module ids" here, e.g. `agent.controlplane.endpoint-manager`. This would be very easy to expose, e.g. have `cell.Module` provide module scoped `type ModuleId string` and `type QualifiedModuleId []ModuleId`. WDYT?

joamaki replied:
https://github.com/cilium/cilium/pull/28512

tommyp1ckles replied:
sounds good, we can fix the merge conflicts on whatever PR is merged second.

tklauser commented on pkg/endpoint/bpf.go:
```diff
@@ -1397,9 +1397,9 @@ func (e *Endpoint) startSyncPolicyMapController() {
+			Group:          syncPolicymapControllerGroup,
+			HealthReporter: e.GetReporter("policy map sync"),
 			DoFunc: func(ctx context.Context) error {
-				// Failure to lock is not an error, it means
```
It looks like this part of the comment was removed by accident?

tommyp1ckles replied:
Hmm I'll fix that.

tommyp1ckles replied:
PR to revert deleted comment line: https://github.com/cilium/cilium/pull/29172
